File: packages/sdk-ts/src/pool.ts
import { CookieNexusClient } from './client.js';

export interface PoolAccount {
  accountId: string;
  vaultId: string;
  domain: string;
  weight?: number;
  lastUsed?: number;
  usageCount: number;
}

export type RotationStrategy = 'round_robin' | 'random' | 'least_used';

export class CookiePool {
  private accounts: PoolAccount[] = [];
  private currentIndex: number = 0;

  constructor(private client: CookieNexusClient) {}

  public addAccount(accountId: string, vaultId: string, domain: string, weight: number = 1) {
    this.accounts.push({
      accountId,
      vaultId,
      domain,
      weight,
      usageCount: 0,
      lastUsed: 0,
    });
  }

  public async acquireSession(strategy: RotationStrategy = 'round_robin'): Promise<{ account: PoolAccount; cookies: any[] }> {
    if (this.accounts.length === 0) {
      throw new Error('No accounts registered in CookiePool');
    }

    let selected: PoolAccount;

    if (strategy === 'random') {
      const idx = Math.floor(Math.random() * this.accounts.length);
      selected = this.accounts[idx];
    } else if (strategy === 'least_used') {
      selected = [...this.accounts].sort((a, b) => a.usageCount - b.usageCount)[0];
    } else {
      // round_robin
      selected = this.accounts[this.currentIndex];
      this.currentIndex = (this.currentIndex + 1) % this.accounts.length;
    }

    selected.usageCount++;
    selected.lastUsed = Date.now();

    const cookies = await this.client.getCookies(selected.domain);
    return { account: selected, cookies };
  }

  public getStats() {
    return this.accounts.map(a => ({
      accountId: a.accountId,
      vaultId: a.vaultId,
      usageCount: a.usageCount,
      lastUsed: a.lastUsed ? new Date(a.lastUsed).toISOString() : 'never',
    }));
  }
}

File: packages/sdk-python/cookienexus/pool.py
import random
import time
from typing import List, Dict, Any, Optional
from .client import CookieNexusClient

class PoolAccount:
    def __init__(self, account_id: str, vault_id: str, domain: str, weight: int = 1):
        self.account_id = account_id
        self.vault_id = vault_id
        self.domain = domain
        self.weight = weight
        self.usage_count = 0
        self.last_used = 0

class CookiePool:
    def __init__(self, client: CookieNexusClient):
        self.client = client
        self.accounts: List[PoolAccount] = []
        self._current_index = 0

    def add_account(self, account_id: str, vault_id: str, domain: str, weight: int = 1) -> None:
        self.accounts.append(PoolAccount(account_id, vault_id, domain, weight))

    def acquire_session(self, strategy: str = "round_robin") -> Dict[str, Any]:
        if not self.accounts:
            raise ValueError("No accounts available in pool")

        if strategy == "random":
            selected = random.choice(self.accounts)
        elif strategy == "least_used":
            selected = min(self.accounts, key=lambda a: a.usage_count)
        else:
            # round_robin
            selected = self.accounts[self._current_index]
            self._current_index = (self._current_index + 1) % len(self.accounts)

        selected.usage_count += 1
        selected.last_used = time.time()

        # Temporarily switch client vault if different
        orig_vault = self.client.vault_id
        self.client.vault_id = selected.vault_id
        try:
            cookies = self.client.get_cookies(selected.domain)
        finally:
            self.client.vault_id = orig_vault

        return {
            "account_id": selected.account_id,
            "vault_id": selected.vault_id,
            "domain": selected.domain,
            "cookies": cookies,
            "cookie_header": "; ".join(f"{c['name']}={c['value']}" for c in cookies)
        }
